File: iosApp/iosApp/Widget/WidgetHost.swift
import SwiftUI
import shared

/// Recursive widget renderer. See ARCHITECTURE Widget/.
struct WidgetHost: View {
    let widget: PageWidget
    let widgetData: Any?
    let onAction: (WidgetAction) -> Void

    var body: some View {
        let context = WidgetRenderContext(widget: widget, widgetData: widgetData, onAction: onAction)
        if let renderer = WidgetRegistry.shared.get(widget.type) {
            renderer(context)
        } else {
            EmptyView()
        }
    }
}

File: androidApp/src/main/kotlin/com/snapp/android/ui/widget/WidgetHost.kt
package com.snapp.android.ui.widget

import androidx.compose.runtime.Composable
import com.snapp.data.model.page.PageWidget

/** Recursive widget renderer. See ARCHITECTURE ui/widget/. */
@Composable
fun WidgetHost(
    widget: PageWidget,
    widgetData: Any? = null,
    onAction: (WidgetAction) -> Unit = {}
) {
    val context = WidgetRenderContext(widget, widgetData, onAction)
    val renderer = WidgetRegistry.get(widget.type)
    renderer?.invoke(context)
}
